contracts/StorageManager.sol

```diff
@@ -242,56 +292,113 @@ contract StorageManager {
+            uint256 maxWithdrawableFunds;
+            if (agreement.lastPayoutDate == 0) {
+                // Agreement is inactive, consumer can withdraw all funds
+                maxWithdrawableFunds = agreement.availableFunds;
+            } else {
+                // Consumer can withdraw all funds except for those already used for past storage hosting
+                // AND for current period
+                maxWithdrawableFunds = agreement.availableFunds.sub(_calculateSpentFunds(agreement)).sub((agreement.billingPrice * agreement.size));
+            }
 
-        if (amount == 0) {
-            amount = maxWithdrawableFunds;
+            if (amount == 0) {
+                amount = maxWithdrawableFunds;
+            }
+            require(amount <= maxWithdrawableFunds, "StorageManager: Amount is too big");
+            agreement.availableFunds = agreement.availableFunds.sub(amount);
+            require(amount > 0, "StorageManager: Nothing to withdraw");
+
+            if(token == address(0)) {
+                (bool success,) = msg.sender.call.value(amount)("");
+                require(success, "Transfer failed.");
+            } else {
+                require(IERC20(token).transfer(msg.sender, amount), "StorageManager: not allowed to deposit tokens from token contract");
+            }
+            emit AgreementFundsWithdrawn(agreementReference, amount, token);
         }
-
-        require(amount <= maxWithdrawableFunds, "StorageManager: Amount is too big");
-        agreement.availableFunds = agreement.availableFunds.sub(amount);
-
-        require(amount > 0, "StorageManager: Nothing to withdraw");
-        (bool success,) = msg.sender.call.value(amount)("");
-        require(success, "Transfer failed.");
-
-        emit AgreementFundsWithdrawn(agreementReference, amount);
     }
 
+
     /**
     >> FOR PROVIDER
     @notice payout already earned funds of one or more Agreement
     @dev
     - Provider must call an expired agreement themselves as soon as the agreement is expired, to add back the capacity to their Offer.
     - Payout can be triggered by other events as well. Like in newAgreement call with either existing agreement or when other
       Agreements are passed to the agreementsReferencesToBePayedOut array.
-    @param agreementReferences reference to one or more Agreement
+    @param dataReferencesOfAgreementToPayOut the data references of the agreement to pay
+    @param creatorsOfAgreementToPayOut the creators that made the agreement to pay
+    @param tokensOfAgreementsToPayOut the tokens of the agreement to pay out
     */
-    function payoutFunds(bytes32[] memory agreementReferences) public {
-        _payoutFunds(agreementReferences, msg.sender);
+
+    function payoutFunds(
+        bytes32[][] memory dataReferencesOfAgreementToPayOut,
+        address[] memory creatorsOfAgreementToPayOut,
+        address tokensOfAgreementsToPayOut,
+        address payable provider
+    ) public whenNotPaused {
```

Comment:
we should allow Withdrawals when Paused.

Reply:
Done